File: node/src/lib/fss.js
/** Pure-JavaScript Forward Secure Pseudorandom Generator (FSPRG).
 *
 * Implements the deterministic key-evolution primitive used by systemd
 * journal Forward Secure Sealing, matching systemd v260.1 behavior.
 */

import { createHash } from 'node:crypto';

export const RECOMMENDED_SECPAR = 1536;
export const RECOMMENDED_SEEDLEN = 12;

function isValidSecpar(secpar) {
  return secpar % 16 === 0 && secpar >= 16 && secpar <= 16384;
}

/** Return the size of the master-secret key for a given secpar. */
export function mskInBytes(secpar) {
  if (!isValidSecpar(secpar)) throw new Error('invalid secpar');
  return 2 + secpar / 8;
}

/** Return the size of the master-public key for a given secpar. */
export function mpkInBytes(secpar) {
  if (!isValidSecpar(secpar)) throw new Error('invalid secpar');
  return 2 + secpar / 8;
}

/** Return the size of an FSPRG state for a given secpar. */
export function stateInBytes(secpar) {
  if (!isValidSecpar(secpar)) throw new Error('invalid secpar');
  return 2 + (2 * secpar) / 8 + 8;
}

function storeSecpar(secpar) {
  const v = secpar / 16 - 1;
  const buf = Buffer.allocUnsafe(2);
  buf.writeUInt16BE(v, 0);
  return buf;
}

function readSecpar(buf) {
  const v = buf.readUInt16BE(0);
  return 16 * (v + 1);
}

function mpiExport(x, buflen) {
  let hex = x.toString(16);
  if (hex.length > buflen * 2) throw new Error('mpiExport: value too large');
  if (hex.length % 2) hex = '0' + hex;
  const out = Buffer.alloc(buflen);
  const b = Buffer.from(hex, 'hex');
  b.copy(out, buflen - b.length);
  return out;
}

function mpiImport(buf) {
  return BigInt('0x' + buf.toString('hex'));
}

function uint64Export(x) {
  const buf = Buffer.allocUnsafe(8);
  buf.writeBigUInt64BE(x, 0);
  return buf;
}

function uint64Import(buf) {
  return buf.readBigUInt64BE(0);
}

/** Deterministically generate buflen pseudorandom bytes from seed and idx. */
function detRandomize(buflen, seed, idx) {
  const out = Buffer.allocUnsafe(buflen);
  let off = 0;
  let ctr = 0;
  const idxBuf = Buffer.allocUnsafe(4);
  idxBuf.writeUInt32BE(idx, 0);
  while (buflen > 0) {
    const h = createHash('sha256');
    h.update(seed);
    h.update(idxBuf);
    const ctrBuf = Buffer.allocUnsafe(4);
    ctrBuf.writeUInt32BE(ctr, 0);
    h.update(ctrBuf);
    const chunk = h.digest();
    const cpylen = Math.min(buflen, 32);
    chunk.copy(out, off, 0, cpylen);
    off += cpylen;
    buflen -= cpylen;
    ctr += 1;
  }
  return out;
}

/** Modular exponentiation: base^exp (mod mod). */
function modPow(base, exp, mod) {
  if (mod === 1n) return 0n;
  let result = 1n;
  let b = base % mod;
  let e = exp;
  while (e > 0n) {
    if (e & 1n) result = (result * b) % mod;
    b = (b * b) % mod;
    e >>= 1n;
  }
  return result;
}

/** Extended Euclidean algorithm returning [g, x, y] such that ax + by = g = gcd(a, b). */
function extendedGcd(a, b) {
  if (b === 0n) return [a, 1n, 0n];
  const [g, x1, y1] = extendedGcd(b, a % b);
  const x = y1;
  const y = x1 - (a / b) * y1;
  return [g, x, y];
}

/** Modular inverse of a (mod m). */
function modInv(a, m) {
  const [g, x] = extendedGcd(a % m, m);
  if (g !== 1n && g !== -1n) throw new Error('modular inverse does not exist');
  return ((x % m) + m) % m;
}

/** Deterministic Miller-Rabin probable-prime test using the first `rounds` prime bases.
 *  Using 64 bases provides a stronger deterministic check than 12 bases and
 *  reduces arbitrary-seed divergence risk.  This does not claim to reproduce
 *  libgcrypt's random witness selection; it is an intentionally stronger
 *  deterministic check.
 */
function isProbablePrime(n, rounds = 64) {
  if (n < 2n) return false;
  if (n === 2n || n === 3n) return true;
  if (n % 2n === 0n) return false;

  let d = n - 1n;
  let r = 0;
  while (d % 2n === 0n) {
    d /= 2n;
    r += 1;
  }

  const bases = [
    2n, 3n, 5n, 7n, 11n, 13n, 17n, 19n, 23n, 29n, 31n, 37n, 41n, 43n, 47n, 53n,
    59n, 61n, 67n, 71n, 73n, 79n, 83n, 89n, 97n, 101n, 103n, 107n, 109n, 113n,
    127n, 131n, 137n, 139n, 149n, 151n, 157n, 163n, 167n, 173n, 179n, 181n,
    191n, 193n, 197n, 199n, 211n, 223n, 227n, 229n, 233n, 239n, 241n, 251n,
    257n, 263n, 269n, 271n, 277n, 281n, 283n, 293n, 307n, 311n,
  ];
  for (let i = 0; i < rounds && i < bases.length; i++) {
    const a = bases[i];
    if (a >= n) continue;
    let x = modPow(a, d, n);
    if (x === 1n || x === n - 1n) continue;
    let cont = false;
    for (let j = 1; j < r; j++) {
      x = modPow(x, 2n, n);
      if (x === n - 1n) {
        cont = true;
        break;
      }
    }
    if (cont) continue;
    return false;
  }
  return true;
}

function genPrime3Mod4(bits, seed, idx) {
  const buflen = bits / 8;
  const buf = Buffer.from(detRandomize(buflen, seed, idx));
  buf[0] |= 0xc0;
  buf[buf.length - 1] |= 0x03;
  let p = mpiImport(buf);
  while (!isProbablePrime(p, 64)) {
    p += 4n;
  }
  return p;
}

function genSquare(n, seed, idx, secpar) {
  const buflen = secpar / 8;
  const buf = Buffer.from(detRandomize(buflen, seed, idx));
  buf[0] &= 0x7f;
  const x = mpiImport(buf);
  if (x >= n) throw new Error('genSquare: x >= n');
  return (x * x) % n;
}

function twopowmodphi(m, p) {
  const phi = p - 1n;
  return modPow(2n, BigInt(m), phi);
}

function crtCompose(xp, xq, p, q) {
  let a = (xq - xp) % q;
  if (a < 0n) a += q;
  const u = modInv(p, q);
  a = (a * u) % q;
  return ((p * a + xp) % (p * q));
}

/** Generate a master key pair deterministically from seed.
 *
 * Returns { msk, mpk } where msk is the master secret key and mpk is the
 * master public key.
 */
export function fsprgGenMK(seed, secpar = RECOMMENDED_SECPAR) {
  if (!isValidSecpar(secpar)) throw new Error('invalid secpar');
  const p = genPrime3Mod4(secpar / 2, seed, 0x01);
  const q = genPrime3Mod4(secpar / 2, seed, 0x02);
  const n = p * q;
  const msk = Buffer.concat([storeSecpar(secpar), mpiExport(p, secpar / 16), mpiExport(q, secpar / 16)]);
  const mpk = Buffer.concat([storeSecpar(secpar), mpiExport(n, secpar / 8)]);
  return { msk, mpk };
}

/** Generate the epoch-0 state from mpk and seed. */
export function fsprgGenState0(mpk, seed) {
  const secpar = readSecpar(mpk);
  const n = mpiImport(mpk.slice(2, 2 + secpar / 8));
  const x = genSquare(n, seed, 0x03, secpar);
  const state = Buffer.alloc(stateInBytes(secpar));
  mpk.copy(state);
  mpiExport(x, secpar / 8).copy(state, 2 + secpar / 8);
  // epoch zero is already zero-initialized
  return state;
}

/** Return the epoch encoded in state. */
export function fsprgGetEpoch(state) {
  const secpar = readSecpar(state);
  return uint64Import(state.slice(2 + 2 * secpar / 8, 2 + 2 * secpar / 8 + 8));
}

/** Evolve state forward by one epoch.  The input is not modified. */
export function fsprgEvolve(state) {
  const secpar = readSecpar(state);
  const n = mpiImport(state.slice(2, 2 + secpar / 8));
  let x = mpiImport(state.slice(2 + secpar / 8, 2 + 2 * secpar / 8));
  let epoch = uint64Import(state.slice(2 + 2 * secpar / 8, 2 + 2 * secpar / 8 + 8));
  x = (x * x) % n;
  epoch += 1n;
  const newState = Buffer.alloc(state.length);
  state.copy(newState);
  mpiExport(x, secpar / 8).copy(newState, 2 + secpar / 8);
  uint64Export(epoch).copy(newState, 2 + 2 * secpar / 8);
  return newState;
}

/** Seek state to an arbitrary epoch using msk and seed.
 *
 * The supplied state must be an epoch-0 state.
 */
export function fsprgSeek(state, epoch, msk, seed) {
  const secpar = readSecpar(msk);
  const p = mpiImport(msk.slice(2, 2 + secpar / 16));
  const q = mpiImport(msk.slice(2 + secpar / 16, 2 + 2 * secpar / 16));
  const n = p * q;
  const x = genSquare(n, seed, 0x03, secpar);
  let xp = x % p;
  let xq = x % q;
  const kp = twopowmodphi(epoch, p);
  const kq = twopowmodphi(epoch, q);
  xp = modPow(xp, kp, p);
  xq = modPow(xq, kq, q);
  const xm = crtCompose(xp, xq, p, q);
  const newState = Buffer.alloc(state.length);
  state.slice(0, 2 + secpar / 8).copy(newState);
  mpiExport(xm, secpar / 8).copy(newState, 2 + secpar / 8);
  uint64Export(BigInt(epoch)).copy(newState, 2 + 2 * secpar / 8);
  return newState;
}

/** Extract a deterministic key from state. */
export function fsprgGetKey(state, keylen, idx) {
  const secpar = readSecpar(state);
  return detRandomize(keylen, state.slice(2, 2 + 2 * secpar / 8 + 8), idx);
}

File: rust/src/crates/journal-core/src/fss.rs
//! Pure-Rust Forward Secure Pseudorandom Generator (FSPRG).
//!
//! Implements the deterministic key-evolution primitive used by systemd
//! journal Forward Secure Sealing, matching systemd v260.1 behavior.

use num_bigint::BigUint;
use sha2::{Digest, Sha256};

pub const RECOMMENDED_SECPAR: u32 = 1536;
pub const RECOMMENDED_SEEDLEN: usize = 12;

fn is_valid_secpar(secpar: u32) -> bool {
    secpar % 16 == 0 && secpar >= 16 && secpar <= 16384
}

/// Return the size of the master-secret key for a given secpar.
pub fn msk_in_bytes(secpar: u32) -> usize {
    assert!(is_valid_secpar(secpar), "invalid secpar");
    2 + (secpar / 8) as usize
}

/// Return the size of the master-public key for a given secpar.
pub fn mpk_in_bytes(secpar: u32) -> usize {
    assert!(is_valid_secpar(secpar), "invalid secpar");
    2 + (secpar / 8) as usize
}

/// Return the size of an FSPRG state for a given secpar.
pub fn state_in_bytes(secpar: u32) -> usize {
    assert!(is_valid_secpar(secpar), "invalid secpar");
    2 + 2 * (secpar / 8) as usize + 8
}

fn store_secpar(secpar: u32) -> [u8; 2] {
    let v = (secpar / 16 - 1) as u16;
    v.to_be_bytes()
}

fn read_secpar(buf: &[u8]) -> u32 {
    let v = u16::from_be_bytes([buf[0], buf[1]]);
    16 * (u32::from(v) + 1)
}

fn mpi_export(x: &BigUint, buflen: usize) -> Vec<u8> {
    let b = x.to_bytes_be();
    assert!(b.len() <= buflen, "mpi_export: value too large for buffer");
    if b.len() == buflen {
        return b;
    }
    let mut out = vec![0u8; buflen];
    out[buflen - b.len()..].copy_from_slice(&b);
    out
}

fn mpi_import(buf: &[u8]) -> BigUint {
    BigUint::from_bytes_be(buf)
}

fn uint64_export(x: u64) -> [u8; 8] {
    x.to_be_bytes()
}

fn uint64_import(buf: &[u8]) -> u64 {
    u64::from_be_bytes([
        buf[0], buf[1], buf[2], buf[3], buf[4], buf[5], buf[6], buf[7],
    ])
}

/// Deterministically generate `buflen` pseudorandom bytes from `seed` and `idx`.
fn det_randomize(buflen: usize, seed: &[u8], idx: u32) -> Vec<u8> {
    let mut out = Vec::with_capacity(buflen);
    // Build the intermediate SHA256 state of seed||idx.
    let mut base = Sha256::new();
    base.update(seed);
    base.update(idx.to_be_bytes());
    // We need to clone the hasher for each counter.  sha2::Sha256 implements Clone.
    let mut ctr: u32 = 0;
    while out.len() < buflen {
        let mut h = base.clone();
        h.update(ctr.to_be_bytes());
        let chunk = h.finalize();
        let cpylen = std::cmp::min(buflen - out.len(), 32);
        out.extend_from_slice(&chunk[..cpylen]);
        ctr += 1;
    }
    out
}

/// Deterministic Miller-Rabin probable-prime test using the first `rounds`
/// prime bases.  Using 64 bases provides a stronger deterministic check than
/// 12 bases and reduces arbitrary-seed divergence risk.  This does not claim
/// to reproduce libgcrypt's random witness selection; it is an intentionally
/// stronger deterministic check.
fn miller_rabin(n: &BigUint, rounds: usize) -> bool {
    if n < &BigUint::from(2u32) {
        return false;
    }
    if *n == BigUint::from(2u32) || *n == BigUint::from(3u32) {
        return true;
    }
    if n.bit(0) == false {
        // even
        return false;
    }

    // Write n-1 as 2^r * d
    let mut d = n - BigUint::from(1u32);
    let mut r = 0;
    while d.bit(0) == false {
        d >>= 1;
        r += 1;
    }

    let bases: [u64; 64] = [
        2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53,
        59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113,
        127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181,
        191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251,
        257, 263, 269, 271, 277, 281, 283, 293, 307, 311,
    ];
    for &a in bases.iter().take(rounds) {
        let a_big = BigUint::from(a);
        if &a_big >= n {
            continue;
        }
        let mut x = a_big.modpow(&d, n);
        let n_minus_1 = n - BigUint::from(1u32);
        if x == BigUint::from(1u32) || x == n_minus_1 {
            continue;
        }
        let mut cont = false;
        for _ in 1..r {
            x = (&x * &x) % n;
            if x == n_minus_1 {
                cont = true;
                break;
            }
        }
        if cont {
            continue;
        }
        return false;
    }
    true
}

fn gen_prime_3mod4(bits: u32, seed: &[u8], idx: u32) -> BigUint {
    let buflen = (bits / 8) as usize;
    let mut buf = det_randomize(buflen, seed, idx);
    buf[0] |= 0xc0;
    let last = buflen - 1;
    buf[last] |= 0x03;
    let four = BigUint::from(4u32);
    let mut p = mpi_import(&buf);
    while !miller_rabin(&p, 64) {
        p += &four;
    }
    p
}

fn gen_square(n: &BigUint, seed: &[u8], idx: u32, secpar: u32) -> BigUint {
    let buflen = (secpar / 8) as usize;
    let mut buf = det_randomize(buflen, seed, idx);
    buf[0] &= 0x7f;
    let x = mpi_import(&buf);
    assert!(x < *n, "genSquare: x >= n");
    (&x * &x) % n
}

fn twopowmodphi(m: u64, p: &BigUint) -> BigUint {
    let phi = p - BigUint::from(1u32);
    BigUint::from(2u32).modpow(&BigUint::from(m), &phi)
}

fn crt_compose(xp: &BigUint, xq: &BigUint, p: &BigUint, q: &BigUint) -> BigUint {
    let mut a = if xq >= xp {
        xq - xp
    } else {
        xq + q - xp
    };
    let u = p
        .modinv(q)
        .expect("CRT: p and q must be coprime");
    a = (&a * &u) % q;
    let n = p * q;
    (p * &a + xp) % &n
}

/// Generate a master key pair deterministically from `seed`.
///
/// Returns `(msk, mpk)` where `msk` is the master secret key and `mpk` is
/// the master public key.
pub fn gen_mk(seed: &[u8], secpar: u32) -> (Vec<u8>, Vec<u8>) {
    assert!(is_valid_secpar(secpar), "invalid secpar");
    let p = gen_prime_3mod4(secpar / 2, seed, 0x01);
    let q = gen_prime_3mod4(secpar / 2, seed, 0x02);
    let n = &p * &q;
    let msk = [
        store_secpar(secpar).as_slice(),
        &mpi_export(&p, (secpar / 16) as usize),
        &mpi_export(&q, (secpar / 16) as usize),
    ]
    .concat();
    let mpk = [
        store_secpar(secpar).as_slice(),
        &mpi_export(&n, (secpar / 8) as usize),
    ]
    .concat();
    (msk, mpk)
}

/// Generate the epoch-0 state from `mpk` and `seed`.
pub fn gen_state0(mpk: &[u8], seed: &[u8]) -> Vec<u8> {
    let secpar = read_secpar(mpk);
    let n = mpi_import(&mpk[2..2 + (secpar / 8) as usize]);
    let x = gen_square(&n, seed, 0x03, secpar);
    let mut state = vec![0u8; state_in_bytes(secpar)];
    state[..mpk.len()].copy_from_slice(mpk);
    state[2 + (secpar / 8) as usize..2 + 2 * (secpar / 8) as usize]
        .copy_from_slice(&mpi_export(&x, (secpar / 8) as usize));
    // epoch zero is already zero-initialized
    state
}

/// Return the epoch encoded in `state`.
pub fn get_epoch(state: &[u8]) -> u64 {
    let secpar = read_secpar(state);
    uint64_import(&state[2 + 2 * (secpar / 8) as usize..2 + 2 * (secpar / 8) as usize + 8])
}

/// Evolve `state` forward by one epoch.  The input is not modified.
pub fn evolve(state: &[u8]) -> Vec<u8> {
    let secpar = read_secpar(state);
    let n = mpi_import(&state[2..2 + (secpar / 8) as usize]);
    let mut x = mpi_import(&state[2 + (secpar / 8) as usize..2 + 2 * (secpar / 8) as usize]);
    let mut epoch = uint64_import(&state[2 + 2 * (secpar / 8) as usize..2 + 2 * (secpar / 8) as usize + 8]);
    x = (&x * &x) % &n;
    epoch += 1;
    let mut new_state = state.to_vec();
    new_state[2 + (secpar / 8) as usize..2 + 2 * (secpar / 8) as usize]
        .copy_from_slice(&mpi_export(&x, (secpar / 8) as usize));
    new_state[2 + 2 * (secpar / 8) as usize..2 + 2 * (secpar / 8) as usize + 8]
        .copy_from_slice(&uint64_export(epoch));
    new_state
}

/// Seek `state` to an arbitrary `epoch` using `msk` and `seed`.
///
/// The supplied `state` must be an epoch-0 state.
pub fn seek(state: &[u8], epoch: u64, msk: &[u8], seed: &[u8]) -> Vec<u8> {
    let secpar = read_secpar(msk);
    let p = mpi_import(&msk[2..2 + (secpar / 16) as usize]);
    let q = mpi_import(&msk[2 + (secpar / 16) as usize..2 + 2 * (secpar / 16) as usize]);
    let n = &p * &q;
    let x = gen_square(&n, seed, 0x03, secpar);
    let mut xp = &x % &p;
    let mut xq = &x % &q;
    let kp = twopowmodphi(epoch, &p);
    let kq = twopowmodphi(epoch, &q);
    xp = xp.modpow(&kp, &p);
    xq = xq.modpow(&kq, &q);
    let xm = crt_compose(&xp, &xq, &p, &q);
    let mut new_state = vec![0u8; state.len()];
    new_state[..2 + (secpar / 8) as usize].copy_from_slice(&state[..2 + (secpar / 8) as usize]);
    new_state[2 + (secpar / 8) as usize..2 + 2 * (secpar / 8) as usize]
        .copy_from_slice(&mpi_export(&xm, (secpar / 8) as usize));
    new_state[2 + 2 * (secpar / 8) as usize..2 + 2 * (secpar / 8) as usize + 8]
        .copy_from_slice(&uint64_export(epoch));
    new_state
}

/// Extract a deterministic key from `state`.
pub fn get_key(state: &[u8], keylen: usize, idx: u32) -> Vec<u8> {
    let secpar = read_secpar(state);
    det_randomize(keylen, &state[2..2 + 2 * (secpar / 8) as usize + 8], idx)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    #[derive(serde::Deserialize)]
    struct VectorKey {
        idx: u32,
        keylen: usize,
        key_hex: String,
    }

    #[derive(serde::Deserialize)]
    struct VectorEpoch {
        epoch: u64,
        state_hex: String,
        seek_state_hex: String,
        keys: Vec<VectorKey>,
    }

    #[derive(serde::Deserialize)]
    struct Vector {
        seed_desc: String,
        seed_hex: String,
        msk_hex: String,
        mpk_hex: String,
        state0_hex: String,
        epochs: Vec<VectorEpoch>,
    }

    #[derive(serde::Deserialize)]
    struct Fixture {
        fsprg_params: serde_json::Value,
        vectors: Vec<Vector>,
    }

    fn load_fixture() -> Fixture {
        let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        // journal-core lives in rust/src/crates/journal-core ; repo root is 4 levels up.
        for _ in 0..4 {
            path.pop();
        }
        path.push("tests");
        path.push("fss");
        path.push("fixtures");
        path.push("fsprg-vectors-v01.json");
        let text = std::fs::read_to_string(&path).unwrap_or_else(|e| {
            panic!("failed to read fixture at {}: {}", path.display(), e)
        });
        serde_json::from_str(&text).expect("invalid fixture JSON")
    }

    #[test]
    fn test_fsprg_vectors() {
        let fixture = load_fixture();
        let secpar = fixture
            .fsprg_params
            .get("secpar")
            .and_then(|v| v.as_u64())
            .expect("missing secpar") as u32;
        assert_eq!(secpar, RECOMMENDED_SECPAR);
        for vec in &fixture.vectors {
            let seed = hex::decode(&vec.seed_hex).expect("decode seed");
            let expected_msk = hex::decode(&vec.msk_hex).expect("decode msk");
            let expected_mpk = hex::decode(&vec.mpk_hex).expect("decode mpk");
            let expected_state0 = hex::decode(&vec.state0_hex).expect("decode state0");
            assert_eq!(seed.len(), RECOMMENDED_SEEDLEN);
            assert_eq!(expected_msk.len(), msk_in_bytes(secpar));
            assert_eq!(expected_mpk.len(), mpk_in_bytes(secpar));

            let (msk, mpk) = gen_mk(&seed, secpar);
            assert_eq!(msk, expected_msk, "msk mismatch for {}", vec.seed_desc);
            assert_eq!(mpk, expected_mpk, "mpk mismatch for {}", vec.seed_desc);

            let state0 = gen_state0(&mpk, &seed);
            assert_eq!(state0, expected_state0, "state0 mismatch for {}", vec.seed_desc);
            assert_eq!(get_epoch(&state0), 0, "epoch0 mismatch for {}", vec.seed_desc);

            for ep in &vec.epochs {
                let mut evolved = state0.clone();
                for _ in 0..ep.epoch {
                    evolved = evolve(&evolved);
                }
                let expected_state = hex::decode(&ep.state_hex).expect("decode state");
                assert_eq!(
                    evolved, expected_state,
                    "evolve mismatch for {} epoch {}",
                    vec.seed_desc, ep.epoch
                );

                let seeked = seek(&state0, ep.epoch, &msk, &seed);
                let expected_seek = hex::decode(&ep.seek_state_hex).expect("decode seek_state");
                assert_eq!(
                    seeked, expected_seek,
                    "seek mismatch for {} epoch {}",
                    vec.seed_desc, ep.epoch
                );

                for k in &ep.keys {
                    let key = get_key(&evolved, k.keylen, k.idx);
                    let expected_key = hex::decode(&k.key_hex).expect("decode key");
                    assert_eq!(
                        key, expected_key,
                        "key mismatch for {} epoch {} idx {}",
                        vec.seed_desc, ep.epoch, k.idx
                    );
                }
            }
        }
    }
}
